src/lib/encoder.ts

```diff
@@ -352,21 +366,90 @@ const valueTypeEncodingMap = {
     },
   },
   uint256: {
-    encode: (value: string | number) =>
-      abiCoder.encodeParameter('uint256', value),
-    decode: (value: string) => abiCoder.decodeParameter('uint256', value),
+    encode: (value: string | number) => {
+      return abiCoder.encodeParameter('uint256', value);
+    },
+    decode: (value: string) => {
+      if (!isHex(value)) {
+        throw new Error(`Can't convert ${value} to uint256, value is not hex.`);
+      }
+
+      const numberOfBytes = countNumberOfBytes(value);
+
+      if (numberOfBytes > 32) {
+        throw new Error(
+          `Can't convert hex value ${value} to uint256. Too many bytes. ${numberOfBytes} is above the maximal number of bytes 32`,
+        );
+      }
+
+      return BigNumber(value).toNumber();
+    },
   },
   bytes32: {
-    encode: (value) => abiCoder.encodeParameter('bytes32', value),
+    encode: (value: string | number) => {
+      let valueToEncode: string;
+
+      if (typeof value === 'string' && !isHex(value)) {
+        // if we receive a plain string (e.g: "hey!"), convert it to utf8-hex data
+        valueToEncode = toHex(value);
+      } else if (typeof value === 'number') {
+        // if we receive a number as input, convert it to hex
+        valueToEncode = numberToHex(value);
+      } else {
+        valueToEncode = value;
+      }
+
+      const numberOfBytes = countNumberOfBytes(valueToEncode);
+
+      // CHECK that there is not more than 4 bytes
+      if (numberOfBytes > 32) {
+        throw new Error(
+          `Can't convert ${value} to bytes32. Too many bytes, expected at most 32 bytes, received ${numberOfBytes}.`,
+        );
+      }
+
+      return abiCoder.encodeParameter('bytes32', valueToEncode);
+    },
     decode: (value: string) => abiCoder.decodeParameter('bytes32', value),
   },
   bytes4: {
-    encode: (value) => abiCoder.encodeParameter('bytes4', value),
-    decode: (value: string) => abiCoder.decodeParameter('bytes4', value),
+    encode: (value: string | number) => {
+      let valueToEncode: string;
+
+      if (typeof value === 'string' && !isHex(value)) {
+        // if we receive a plain string (e.g: "hey!"), convert it to utf8-hex data
+        valueToEncode = toHex(value);
+      } else if (typeof value === 'number') {
+        // if we receive a number as input, convert it to hex
+        valueToEncode = numberToHex(value);
+      } else {
+        valueToEncode = value;
+      }
+
+      const numberOfBytes = countNumberOfBytes(valueToEncode);
+
+      // CHECK that there is not more than 32 bytes
+      if (numberOfBytes > 4) {
+        throw new Error(
+          `Can't convert ${value} to bytes4. Too many bytes, expected at most 4 bytes, received ${numberOfBytes}.`,
+        );
+      }
+
+      const abiEncodedValue = abiCoder.encodeParameter('bytes4', valueToEncode);
+      const bytesArray = hexToBytes(abiEncodedValue);
+      return bytesToHex(bytesArray.slice(0, 4));
```

Comment:
This is duplicated code with L390-L411.

I suggest you to make another function with an input param like `encodeTo: 'bytes32' | 'bytes4'` or maybe an int input, and then you reuse this function form both places

Reply:
DRY

Reply:
I will got with the value type input, as number should be only between 1 and 32.